onFlush and postFlush Entity Listeners

alexgurrola opened the pull request:
Simply check if an onFlush or postFlush entity listener is subscribed
for the current class in the entity manager during dispatch.

alexgurrola committed:
onFlush and postFlush Entity Listeners

Simply check if an onFlush or postFlush entity listener is subscribed
for the current class in the entity manager during dispatch.

Ocramius commented:
Seems like a follow-up of #5582 ?

Ocramius commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3310,13 +3310,49 @@ private function dispatchOnFlushEvent()
         if ($this->evm->hasListeners(Events::onFlush)) {
             $this->evm->dispatchEvent(Events::onFlush, new OnFlushEventArgs($this->em));
         }
+        foreach ($this->entityChangeSets as $oid => $entityChangeSet) {
```
The subscribed listeners can be fetched once the class name is known. No need to fetch a listener for each single entity.

alexgurrola commented:
@Ocramius Correct, with the added postFlush and former commits wrapped together.

mwansinck commented:
Any progress on this?

alexgurrola commented:
@mwansinck The logic worked, but was not optimized.  I could attempt to golf it a bit more this weekend and see if @Ocramius approves of those changes.

mwansinck commented:
@alexgurrola Thanks for your response, I hope your pull request will be accepted. If I can help you in any way, let me know,

mwansinck commented:
@alexgurrola Any update on this?

alexgurrola committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	lib/Doctrine/ORM/Mapping/Builder/EntityListenerBuilder.php

alexgurrola committed:
Merge remote-tracking branch 'upstream/master'

alexgurrola committed:
Optimized OnFlush & PostFlush Listener Invocation

alexgurrola commented:
@mwansinck Thanks much.  I had a little extra time, so I fixed the merge conflicts and added a simple map for the listener subscriptions, to reduce the amount of fetches.  I hope that helps it get approved.  What say you, @Ocramius?

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
+            foreach ($this->identityMap as $className => $entities) {
+                $hash = (isset($this->entityIdentifiers[$oid]) && isset($this->entityIdentifiers[$oid]['id'])) ? $this->entityIdentifiers[$oid]['id'] : null;
+                if ($hash) {
+                    if (is_object($hash) && method_exists($hash, '__toString')) $hash = $hash->__toString();
```
```php
$hash = (string) $hash;
```

Is a bit cleaner than this `if`, don't you think?

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
         if ($this->evm->hasListeners(Events::postFlush)) {
             $this->evm->dispatchEvent(Events::postFlush, new PostFlushEventArgs($this->em));
         }
+        $subscribedSystems = [];
```
This whole part can be extracted to a method instead of duplicated with minor changes.

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3299,13 +3299,59 @@ private function dispatchOnFlushEvent()
             $this->evm->dispatchEvent(Events::onFlush, new OnFlushEventArgs($this->em));
         }
+        $subscribedSystems = [];
+        foreach ($this->entityChangeSets as $oid => $entityChangeSet) {
```
Maybe refactoring (extract method, mostly) this to became more readable?

lcobucci replied:
@Ocramius this nested `foreach` would have a (small) performance impact, should we mind that?

lcobucci commented:
@alexgurrola could you also `rebase` the branch instead of `merging`? And we do need tests 😄
